src/tbb/scheduler_common.h

```diff
@@ -439,6 +440,58 @@ struct suspend_point_type {
 #pragma warning( disable: 4324 )
 #endif
 
+class thread_reference_vertex : public d1::wait_tree_vertex_interface {
+public:
+    thread_reference_vertex(wait_tree_vertex_interface *parent, std::uint32_t ref_count)
+        : m_parent{parent}, m_ref_count{ref_count} {}
+
+    void reserve(std::uint32_t delta = 1) override {
+        auto ref = m_ref_count.fetch_add(static_cast<std::uint64_t>(delta));
+        if (ref == 0) {
+            m_parent->reserve();
+        }
+    }
+
+    void release(std::uint32_t delta = 1) override {
+        // Saving the pointer to the parent before decrementing the reference count
+        // since it won't be safe access any members after that
```

Comment:
I would add more details about the nature of unsafe access:

```suggestion
        // Saving the pointer to the parent before decrementing the reference count
        // since other thread can destroy the vertex after the decrement
```

Feel free to change the wording:)